### ISSUE TITLE
Unexpected behavior when using () -> T to break a cycle

### ISSUE THREAD

**Issue by damianw** [evant/kotlin-inject#286]
If a provision has `@Assisted` injection parameters and injects a factory function for its own type, the code generated for the factory function unexpectedly ignores the assisted arguments and returns a single instance. For example:

```Kotlin
@Inject
class MyDep(
    private val factory: (id: Int) -> MyDep,
    @Assisted val id: Int
) {

    fun next(): MyDep = factory(id + 1)
}

@Component
interface MyComponent {

    val myDepFactory: (id: Int) -> MyDep
}
```

Produces the following code:

```Kotlin
public fun KClass<MyComponent>.create(): MyComponent = InjectMyComponent()

public class InjectMyComponent : MyComponent {
  public override val myDepFactory: Function1<Int, MyDep>
    get() = { arg0 ->
      run {
        lateinit var myDep: MyDep
        MyDep(
          factory = { arg0_ ->
            myDep
          },
          id = arg0
        ).also {
          myDep = it
        }
      }

    }
}
```

Note that `arg0_` is ignored, and the same instance is always returned. In this example, calling `next()` would return the same instance it was called on. This happens with `@Provides` in addition to `@Inject`.

**Comment by evant** [comment]
Hm, on first glance seems like this is a hole in the loop detection and should cause an error to be thrown. If you don't think it is could you write what code you'd expect to generate? I'm having trouble wrapping my head around it.

**Comment by evant** [comment]
Update: this would be possible by generating the code below:
```kotlin
public class InjectMyComponent : MyComponent {
    public override val myDepFactory: (id: Int) -> MyDep
        get() = object : Function1<Int, MyDep> {
            override fun invoke(arg0: Int): MyDep {
                return MyDep(this, arg0) 
            }
        }
}
```
however, this is non-trivial to do. So I'm going to update it to reject assisted injection in cycles and repurpose this ticket to maybe add it in a future release.

**Comment by damianw** [comment]
@evant Thanks for looking into it, and apologies for the late reply. That seems like a reasonable workaround.

I found however that this problem isn't exclusive to `Assisted` injection. For example, the following code:

```Kotlin
@Inject
class MyDep(
    private val factory: () -> MyDep,
) {

    fun next(): MyDep = factory()
}

@Component
interface MyComponent {

    val myDepFactory: () -> MyDep
}
```

Also produces unexpected results:

```Kotlin
public fun KClass<MyComponent>.create(): MyComponent = InjectMyComponent()

public class InjectMyComponent : MyComponent {
  public override val myDepFactory: Function0<MyDep>
    get() = {
      run {
        lateinit var myDep: MyDep
        MyDep(
          factory = {
            myDep
          }
        ).also {
          myDep = it
        }
      }

    }
}
```

This case is perhaps a little less cut-and-dry than the `Assisted` case, but is still kind of unexpected. I would have expected that injecting a `() -> MyDep` would be the equivalent to the Dagger pattern of injecting a `Provider<MyDep>`, but that is is not the case.

**Comment by evant** [comment]
Yes both `Lazy<T>` and `() -> T` break cycles in more or less the same way to give you a circular reference. I agree that this can be surprising and probably should be changed.

**Comment by damianw** [comment]
Gotcha, makes sense. Yeah, arguably the dependency of `() -> T` is a different dependency than `T` and thus shouldn't participate in the cycle breaking, since it's not really a cycle. The most surprising part of this behavior is that the implementation of `() -> T` injected is different depending on whether it's part of a cycle or not. As it is, it's impossible to inject a factory for an injected type without risking running into this problem.

**Comment by sergeshustoff** [comment]
@evant are you ok with the idea that Lazy and lambdas would always create new instances (unless things are scoped) instead of reusing existing instances from the existing cycle-breaking mechanism? I suppose such change should be done under a config flag for gradual migration.

Or here is a crazy idea: always fail on any cycles, unless lambda/lazy dependency that breaks the cycle is marked with either `@CyclicFactory` or `@CyclicReference`? It's much more difficult path with supporting both cycle-breaking mechanisms at once, but quite versatile... And I have some ideas on how to solve problems for the old mechanism (that would work under `@CyclicReference`). Or maybe we could do the same withouot extra annotations - Lazy always reuses instance in a cycle and lambda always creates new one? But that's not quite as explicit as I'd like.

To handle old cycle breaking we could in theory get the whole cycle, find it's lexigraphically lowest class, use it as a kind of key for the whole cycle, create a factory that builds the whole cycle and stores each node in something like scope's lazy map. Then when we need any dependency from such cycle we just use that factory and get it from the resulting map? Complicated, but it might just work).

Either way for the new cycle beaking mechanism, besides creating objects implementing functions as you suggested above, there is a way without objects:

    class CyclicThing(val factory: () -> CyclicThing)
    val factory: () -> CyclicThing = {
        CyclicThing(::factory.invoke())
    }

Thogh such factories have to be extracted as separate members when not declared explicidly as abstract in component, like what happens with optimizatinons that extract frequently used things into private properties in generated code.

Anyway, that's a lot of tricky stuff, so how would you want to proceed with this? Would you want to keep and improve old mechanism that reuses created instances (then I'd assume it's better to do that first) or to replace it with new one entirely?
Would you prefer explicit annotations for cycle breaking or just do it as it's done now implicitly?
